### PR TITLE
Strongly-typed `state.value` and `state.matches`

### PR DESCRIPTION
**Pull request by davidkpiano**
This only works with machines from `setup(...).createMachine(...)`

### CODE REVIEW

File: packages/core/src/State.ts [hunk 1]
```diff
@@ -39,7 +41,8 @@ interface MachineSnapshotBase<
   TChildren extends Record<string, AnyActorRef | undefined>,
   TTag extends string,
   TOutput,
-  TResolvedTypesMeta = TypegenDisabled
+  TResolvedTypesMeta = TypegenDisabled,
+  TConfig extends StateConfig2 = any
```
Comment (Andarist) [inline packages/core/src/State.ts:45]: this shouldn't receive the whole `TConfig` but rather the `TStateValue` - in other words, we should precompute this and pass the resolved value here

File: packages/core/src/types.ts [hunk 2]
```diff
@@ -2368,3 +2354,40 @@ export type ToChildren<TActor extends ProvidedActor> =
...
+    ?
+        | keyof TConfig['states']
+        | {
+            [K in keyof TConfig['states']]?: StateValueFromConfig<
+              TConfig['states'][K]
+            >;
+          }
+    : never;
+
+export type StateValueFrom2<T extends StateConfig2> = IsAny<T> extends true
```
Comment (Andarist) [inline packages/core/src/types.ts:2375]: it would be *very* nice to rewrite this as a tail-recursive conditional type - but we can optimize this later on

File: packages/core/src/types.ts [hunk 3]
```diff
@@ -2368,3 +2354,40 @@ export type ToChildren<TActor extends ProvidedActor> =
...
               : 'exclude']
       >;
+
+export type StateConfig2 = {
+  states?: {
+    [K: string]: StateConfig2;
+  };
+};
+
+export type StateValueFromConfig<TConfig extends StateConfig2> =
```
Comment (Andarist) [inline packages/core/src/types.ts:2364]: this stays unused so it can probably be removed

File: packages/core/src/StateMachine.ts [hunk 4]
```diff
@@ -83,7 +84,8 @@ export class StateMachine<
     TGuard,
     TDelay,
     TTag
-  >
+  >,
+  TConfig extends StateConfig2 = any
```
Comment (Andarist) [inline packages/core/src/StateMachine.ts:88]: We can't infer it here at the moment so it shouldn't even be added to the type parameters list. If we stop passing `TConfig` down (like suggested in another comment) then we could just pass down directly `StateConfig2` (or similar) as `TStateValue`.

File: packages/core/src/State.ts [hunk 5]
```diff
@@ -235,39 +247,44 @@ export type MachineSnapshot<
...
```
Comment (Andarist) [inline packages/core/src/State.ts:251]: Please keep `TResolvedTypesMeta` as the last type parameter in the list

File: packages/core/src/types.ts [hunk 6]
```diff
@@ -2173,25 +2173,11 @@ export type SnapshotFrom<T> = ReturnTypeOrValue<T> extends infer R
     ? TSnapshot
     : R extends Actor<infer TLogic>
       ? SnapshotFrom<TLogic>
-      : R extends StateMachine<
```
Comment (Andarist) [inline packages/core/src/types.ts:2176]: There was a reason why this had two separate branches for `StateMachine` and `ActorLogic`, although I don't recall what that was right now. I'd probably prefer to leave this change for another PR

**Review by davidkpiano — COMMENTED** [review]
LGTM, merge when ready
